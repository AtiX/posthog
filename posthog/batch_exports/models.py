from dataclasses import dataclass
from uuid import UUID
from django.db import models, transaction

from asgiref.sync import sync_to_async
from posthog.models.utils import UUIDModel


class BatchExportDestination(UUIDModel):
    """
    A model for the destination that a PostHog BatchExport will target.

    This model answers the question: where are we exporting data? It contains
    all the necessary information to interact with a specific destination. As we
    wish to support multiple destinations, this forces us to relax schema
    requirements for any configuration parameters, as different destinations
    will have different configuration parameters.
    """

    class Destination(models.TextChoices):
        """Enumeration of supported destinations for PostHog BatchExports."""

        S3 = "S3"
        SNOWFLAKE = "Snowflake"

    secret_fields = {
        "S3": {"aws_access_key_id", "aws_secret_access_key"},
        "Snowflake": set("password"),
    }

    type: models.CharField = models.CharField(
        choices=Destination.choices, max_length=64, help_text="A choice of supported BatchExportDestination types."
    )
    config: models.JSONField = models.JSONField(
        default=dict,
        blank=True,
        help_text="A JSON field to store all configuration parameters required to access a BatchExportDestination.",
    )
    created_at: models.DateTimeField = models.DateTimeField(
        auto_now_add=True, help_text="The timestamp at which this BatchExportDestination was created."
    )
    last_updated_at: models.DateTimeField = models.DateTimeField(
        auto_now=True, help_text="The timestamp at which this BatchExportDestination was last updated."
    )


class BatchExportRun(UUIDModel):
    """
    A model representing a single run of a PostHog BatchExport given a time
    interval. It is used to keep track of the status and progress of the export
    between the specified time interval, as well as communicating any errors
    that may have occurred during the process.
    """

    class Status(models.TextChoices):
        """Possible states of the BatchExportRun."""

        CANCELLED = "Cancelled"
        COMPLETED = "Completed"
        CONTINUEDASNEW = "ContinuedAsNew"
        FAILED = "Failed"
        TERMINATED = "Terminated"
        TIMEDOUT = "TimedOut"
        RUNNING = "Running"
        STARTING = "Starting"

    batch_export = models.ForeignKey(
        "BatchExport", on_delete=models.CASCADE, help_text="The BatchExport this run belongs to."
    )
    status: models.CharField = models.CharField(
        choices=Status.choices, max_length=64, help_text="The status of this run."
    )
    records_completed: models.IntegerField = models.IntegerField(
        null=True, help_text="The number of records that have been exported."
    )
    latest_error: models.TextField = models.TextField(
        null=True, help_text="The latest error that occurred during this run."
    )
    data_interval_start: models.DateTimeField = models.DateTimeField(help_text="The start of the data interval.")
    data_interval_end: models.DateTimeField = models.DateTimeField(help_text="The end of the data interval.")
    cursor: models.TextField = models.TextField(null=True, help_text="An opaque cursor that may be used to resume.")
    created_at: models.DateTimeField = models.DateTimeField(
        auto_now_add=True, help_text="The timestamp at which this BatchExportRun was created."
    )
    finished_at: models.DateTimeField = models.DateTimeField(
        null=True, help_text="The timestamp at which this BatchExportRun finished, successfully or not."
    )
    last_updated_at: models.DateTimeField = models.DateTimeField(
        auto_now=True, help_text="The timestamp at which this BatchExportRun was last updated."
    )


class BatchExport(UUIDModel):
    """
    Defines the configuration of PostHog to export data to a destination,
    either on a schedule (via the interval parameter), or manually by a
    "backfill". Specific instances of a unit process of exporting data is called
    a BatchExportRun.
    """

    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE, help_text="The team this belongs to.")
    name: models.TextField = models.TextField(help_text="A human-readable name for this BatchExport.")
    destination: models.ForeignKey = models.ForeignKey(
        "BatchExportDestination", on_delete=models.CASCADE, help_text="The destination to export data to."
    )
    interval = models.CharField(
        max_length=64,
        null=False,
        choices=[("hour", "hour"), ("day", "day"), ("week", "week")],
        default="hour",
        help_text="The interval at which to export data.",
    )
    paused = models.BooleanField(default=False, help_text="Whether this BatchExport is paused or not.")
    deleted = models.BooleanField(default=False, help_text="Whether this BatchExport is deleted or not.")
    created_at: models.DateTimeField = models.DateTimeField(
        auto_now_add=True, help_text="The timestamp at which this BatchExport was created."
    )
    last_updated_at: models.DateTimeField = models.DateTimeField(
        auto_now=True, help_text="The timestamp at which this BatchExport was last updated."
    )


@dataclass
class BatchExportDestinationData:
    """
    Static structures that we can easily pass around to, e.g. asyncio tasks.
    """

    type: str
    config: dict


@dataclass
class BatchExportData:
    """
    Static structures that we can easily pass around to, e.g. asyncio tasks.
    """

    id: UUID
    team_id: int
    name: str
    interval: str
    destination: BatchExportDestinationData


def acreate_batch_export(
    team_id: int, name: str, type: str, config: dict, interval: str, paused: bool = False
) -> BatchExportData:
    """
    Creates a BatchExport for a given team, destination, and interval.

    :param team: The team to create the BatchExport for.
    :param name: The name of the BatchExport.
    :param destination: The destination to export data to.
    :param interval: The interval at which to export data.
    :param paused: Whether the BatchExport should be paused or not.
    :return: The created BatchExport.
    """
    with transaction.atomic():
        destination = BatchExportDestination.objects.create(type=type, config=config)
        export = BatchExport.objects.create(
            team_id=team_id, name=name, destination=destination, interval=interval, paused=paused
        )

    return BatchExportData(
        id=export.id,
        team_id=export.team_id,
        name=export.name,
        interval=export.interval,
        destination=BatchExportDestinationData(
            type=export.destination.type,
            config=export.destination.config,
        ),
    )


def fetch_batch_export(batch_export_id: UUID) -> BatchExport | None:
    """
    Fetch a BatchExport by id.
    """
    try:
        export_row = BatchExport.objects.values(
            "id", "team_id", "name", "interval", "destination__type", "destination__config"
        ).get(id=batch_export_id)
    except BatchExport.DoesNotExist:
        return None

    return BatchExport(
        id=export_row["id"],
        team_id=export_row["team_id"],
        name=export_row["name"],
        interval=export_row["interval"],
        destination=BatchExportDestination(
            type=export_row["destination__type"],
            config=export_row["destination__config"],
        ),
    )


async def afetch_batch_export(batch_export_id: UUID) -> BatchExport | None:
    """
    Fetch a BatchExport by id.
    """
    return await sync_to_async(fetch_batch_export)(batch_export_id)  # type: ignore
