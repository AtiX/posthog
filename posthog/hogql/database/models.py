from typing import Any, Callable, Dict, List, Optional, TYPE_CHECKING
from pydantic import BaseModel, Extra

from posthog.hogql.errors import HogQLException, NotImplementedException

if TYPE_CHECKING:
    from posthog.hogql.context import HogQLContext
    from posthog.hogql.base import ConstantType


class FieldOrTable(BaseModel):
    pass


class DatabaseField(FieldOrTable):
    """
    Base class for a field in a database table.
    """

    class Config:
        extra = Extra.forbid

    name: str
    array: Optional[bool] = None
    nullable: Optional[bool] = None

    def is_nullable(self) -> bool:
        return not not self.nullable

    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import UnknownType

        return UnknownType(nullable=self.is_nullable())


class IntegerDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import FloatType

        return FloatType(nullable=self.is_nullable())


class FloatDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import IntegerType

        return IntegerType(nullable=self.is_nullable())


class StringDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import StringType

        return StringType(nullable=self.is_nullable())


class StringJSONDatabaseField(DatabaseField):
    pass


class StringArrayDatabaseField(DatabaseField):
    pass


class DateDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import DateType

        return DateType(nullable=self.is_nullable())


class DateTimeDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import DateTimeType

        return DateTimeType(nullable=self.is_nullable())


class BooleanDatabaseField(DatabaseField):
    def get_constant_type(self) -> "ConstantType":
        from posthog.hogql.ast import BooleanType

        return BooleanType(nullable=self.is_nullable())


class FieldTraverser(FieldOrTable):
    class Config:
        extra = Extra.forbid

    chain: List[str]


class Table(FieldOrTable):
    fields: Dict[str, FieldOrTable]

    class Config:
        extra = Extra.forbid

    def has_field(self, name: str) -> bool:
        return name in self.fields

    def get_field(self, name: str) -> FieldOrTable:
        if self.has_field(name):
            return self.fields[name]
        raise Exception(f'Field "{name}" not found on table {self.__class__.__name__}')

    def to_printed_clickhouse(self, context: "HogQLContext") -> str:
        raise NotImplementedException("Table.to_printed_clickhouse not overridden")

    def to_printed_hogql(self) -> str:
        raise NotImplementedException("Table.to_printed_hogql not overridden")

    def avoid_asterisk_fields(self) -> List[str]:
        return []

    def get_asterisk(self):
        fields_to_avoid = self.avoid_asterisk_fields() + ["team_id"]
        asterisk: Dict[str, FieldOrTable] = {}
        for key, field in self.fields.items():
            if key in fields_to_avoid:
                continue
            if isinstance(field, DatabaseField):
                asterisk[key] = field
            elif isinstance(field, Table) or isinstance(field, LazyJoin) or isinstance(field, FieldTraverser):
                pass  # ignore virtual tables for now
            else:
                raise HogQLException(f"Unknown field type {type(field).__name__} for asterisk")
        return asterisk


class LazyJoin(FieldOrTable):
    class Config:
        extra = Extra.forbid

    join_function: Callable[[str, str, Dict[str, Any]], Any]
    join_table: Table
    from_field: str


class LazyTable(Table):
    """
    A table that is replaced with a subquery returned from `lazy_select(requested_fields: Dict[name, chain])`
    """

    class Config:
        extra = Extra.forbid

    def lazy_select(self, requested_fields: Dict[str, List[str]]) -> Any:
        raise NotImplementedException("LazyTable.lazy_select not overridden")


class VirtualTable(Table):
    """
    A nested table that reuses the parent for storage. E.g. events.person.* fields with PoE enabled.
    """

    class Config:
        extra = Extra.forbid


class FunctionCallTable(Table):
    """
    A table that returns a function call, e.g. numbers(...) or s3(...). The team_id guard is NOT added for these.
    """

    name: str
    min_args: Optional[int] = None
    max_args: Optional[int] = None


class SavedQuery(Table):
    """
    A table that returns a subquery, e.g. my_saved_query -> (SELECT * FROM some_saved_table). The team_id guard is NOT added for the overall subquery
    """

    query: str
    name: str

    # Note: redundancy for safety. This validation is used in the data model already
    def to_printed_clickhouse(self, context):
        from posthog.warehouse.models import validate_saved_query_name

        validate_saved_query_name(self.name)
        return self.name

    def to_printed_hogql(self):
        from posthog.warehouse.models import validate_saved_query_name

        validate_saved_query_name(self.name)
        return self.name
