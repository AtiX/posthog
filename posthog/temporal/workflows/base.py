from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from typing import Any
from uuid import UUID

from posthog.batch_exports.service import afetch_batch_export, update_batch_export_run_status, create_batch_export_run
from temporalio import activity
from asgiref.sync import sync_to_async

from posthog.temporal.workflows.batch_exports import get_data_interval_from_workflow_inputs


class PostHogWorkflow(ABC):
    """Base class for Temporal Workflows that can be executed in PostHog."""

    @classmethod
    def get_name(cls) -> str:
        """Get this workflow's name."""
        return getattr(cls, "__temporal_workflow_definition").name

    @classmethod
    def is_named(cls, name: str) -> bool:
        """Check if this workflow's name matches name.

        All temporal workflows have the __temporal_workflow_definition attribute
        injected into them by the defn decorator. We use it to access the name and
        avoid having to define it twice. If this changes in the future, we can
        update this method instead of changing every single workflow.
        """
        return cls.get_name() == name

    @staticmethod
    @abstractmethod
    def parse_inputs(inputs: list[str]) -> Any:
        """Parse inputs from the management command CLI.

        If a workflow is to be executed via the CLI it must know how to parse its
        own inputs.
        """
        return NotImplemented


@dataclass
class CreateBatchExportRunInputs:
    """Inputs to the create_export_run activity.

    Attributes:
        batch_export_id: The ID of the BatchExport that this run is for.
        scheduled_start_time: The time at which the export was scheduled to
        start. If it's None we assume that it.
    """

    batch_export_id: str
    scheduled_start_time: str


@activity.defn
async def create_export_run(inputs: CreateBatchExportRunInputs) -> str | None:
    """Activity that creates an BatchExportRun.

    Intended to be used in all export workflows, usually at the start, to create a model
    instance to represent them in our database.
    """
    activity.logger.info("Creating BatchExportRun model for export: %s", inputs.batch_export_id)

    batch_export = await afetch_batch_export(UUID(inputs.batch_export_id))
    if not batch_export:
        activity.logger.warn(f"BatchExport {inputs.batch_export_id} does not exist")
        return

    datetime.fromisoformat(inputs.scheduled_start_time)
    (data_interval_start, data_interval_end) = get_data_interval_from_workflow_inputs(
        interval="hour", data_interval_end=datetime.fromisoformat(inputs.scheduled_start_time)
    )

    # 'sync_to_async' type hints are fixed in asgiref>=3.4.1
    # But one of our dependencies is pinned to asgiref==3.3.2.
    # Remove these comments once we upgrade.
    run = await sync_to_async(create_batch_export_run)(  # type: ignore
        batch_export_id=UUID(inputs.batch_export_id),
        data_interval_start=data_interval_start,
        data_interval_end=data_interval_end,
    )

    activity.logger.info(f"Created BatchExportRun {run.id}")

    return str(run.id)


@dataclass
class UpdateBatchExportRunStatusInputs:
    """Inputs to the update_export_run_status activity."""

    id: str
    status: str


@activity.defn
async def update_export_run_status(inputs: UpdateBatchExportRunStatusInputs):
    """Activity that updates the status of an BatchExportRun."""
    await sync_to_async(update_batch_export_run_status)(run_id=UUID(inputs.id), status=inputs.status)  # type: ignore
