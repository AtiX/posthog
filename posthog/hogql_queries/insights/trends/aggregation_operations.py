from typing import List
from posthog.hogql import ast
from posthog.hogql.parser import parse_expr, parse_select
from posthog.hogql_queries.utils.query_date_range import QueryDateRange
from posthog.schema import ActionsNode, EventsNode


class QueryAlternator:
    """Allows query_builder to modify the query without having to expost the whole AST interface"""

    _query: ast.SelectQuery
    _selects: List[ast.Expr]
    _group_bys: List[ast.Expr]
    _select_from: ast.JoinExpr | None

    def __init__(self, query: ast.SelectQuery | ast.SelectUnionQuery):
        assert isinstance(query, ast.SelectQuery)

        self._query = query
        self._selects = []
        self._group_bys = []
        self._select_from = None

    def build(self) -> ast.SelectQuery | ast.SelectUnionQuery:
        if len(self._selects) > 0:
            self._query.select.extend(self._selects)

        if len(self._group_bys) > 0:
            if self._query.group_by is None:
                self._query.group_by = self._group_bys
            else:
                self._query.group_by.extend(self._group_bys)

        if self._select_from is not None:
            self._query.select_from = self._select_from

        return self._query

    def append_select(self, expr: ast.Expr) -> None:
        self._selects.append(expr)

    def append_group_by(self, expr: ast.Expr) -> None:
        self._group_bys.append(expr)

    def replace_select_from(self, join_expr: ast.JoinExpr) -> None:
        self._select_from = join_expr


class AggregationOperations:
    series: EventsNode | ActionsNode
    query_date_range: QueryDateRange

    def __init__(self, series: EventsNode | ActionsNode, query_date_range: QueryDateRange) -> None:
        self.series = series
        self.query_date_range = query_date_range

    def select_aggregation(self) -> ast.Expr:
        if self.series.math == "hogql" and self.series.math_hogql is not None:
            return parse_expr(self.series.math_hogql)
        elif self.series.math == "total":
            return parse_expr("count(e.uuid)")
        elif self.series.math == "dau":
            return parse_expr("count(DISTINCT e.person_id)")
        elif self.series.math == "weekly_active":
            return ast.Field(chain=["counts"])  # This gets replaced when doing query orchestration
        elif self.series.math == "monthly_active":
            return ast.Field(chain=["counts"])  # This gets replaced when doing query orchestration
        elif self.series.math == "unique_session":
            return parse_expr('count(DISTINCT e."$session_id")')
        elif self.series.math == "unique_group" and self.series.math_group_type_index is not None:
            return parse_expr(f'count(DISTINCT e."$group_{self.series.math_group_type_index}")')
        elif self.series.math_property is not None:
            if self.series.math == "avg":
                return self._math_func("avg")
            elif self.series.math == "sum":
                return self._math_func("sum")
            elif self.series.math == "min":
                return self._math_func("min")
            elif self.series.math == "max":
                return self._math_func("max")
            elif self.series.math == "median":
                return self._math_func("median")
            elif self.series.math == "p90":
                return self._math_quantile(0.9)
            elif self.series.math == "p95":
                return self._math_quantile(0.95)
            elif self.series.math == "p99":
                return self._math_quantile(0.99)
            else:
                raise NotImplementedError()

        return parse_expr("count(e.uuid)")

    def requires_query_orchestration(self) -> bool:
        return self.series.math == "weekly_active" or self.series.math == "monthly_active"

    def _math_func(self, method: str) -> ast.Call:
        if self.series.math_property == "$time":
            return ast.Call(
                name=method,
                args=[
                    ast.Call(
                        name="toUnixTimestamp",
                        args=[ast.Field(chain=["properties", "$time"])],
                    )
                ],
            )

        if self.series.math_property == "$session_duration":
            chain = ["session", "duration"]
        else:
            chain = ["properties", self.series.math_property]
        return ast.Call(name=method, args=[ast.Field(chain=chain)])

    def _math_quantile(self, percentile: float) -> ast.Call:
        return ast.Call(
            name="quantile",
            params=[ast.Constant(value=percentile)],
            args=[ast.Field(chain=["properties", self.series.math_property])],
        )

    def _interval_placeholders(self):
        if self.series.math == "weekly_active":
            return {
                "exclusive_lookback": ast.Call(name="toIntervalDay", args=[ast.Constant(value=6)]),
                "inclusive_lookback": ast.Call(name="toIntervalDay", args=[ast.Constant(value=7)]),
            }
        elif self.series.math == "monthly_active":
            return {
                "exclusive_lookback": ast.Call(name="toIntervalDay", args=[ast.Constant(value=29)]),
                "inclusive_lookback": ast.Call(name="toIntervalDay", args=[ast.Constant(value=30)]),
            }

        raise NotImplementedError()

    def _parent_select_query(
        self, inner_query: ast.SelectQuery | ast.SelectUnionQuery
    ) -> ast.SelectQuery | ast.SelectUnionQuery:
        return parse_select(
            """
                SELECT
                    counts AS total,
                    dateTrunc({interval}, timestamp) AS day_start
                FROM {inner_query}
                WHERE timestamp >= {date_from} AND timestamp <= {date_to}
            """,
            placeholders={
                **self.query_date_range.to_placeholders(),
                "inner_query": inner_query,
            },
        )

    def _inner_select_query(
        self, cross_join_select_query: ast.SelectQuery | ast.SelectUnionQuery
    ) -> ast.SelectQuery | ast.SelectUnionQuery:
        return parse_select(
            """
                SELECT
                    d.timestamp,
                    COUNT(DISTINCT actor_id) AS counts
                FROM (
                    SELECT
                        toStartOfDay({date_to}) - toIntervalDay(number) AS timestamp
                    FROM
                        numbers(dateDiff('day', toStartOfDay({date_from} - {inclusive_lookback}), {date_to}))
                ) d
                CROSS JOIN {cross_join_select_query} e
                WHERE
                    e.timestamp <= d.timestamp + INTERVAL 1 DAY AND
                    e.timestamp > d.timestamp - {exclusive_lookback}
                GROUP BY d.timestamp
                ORDER BY d.timestamp
            """,
            placeholders={
                **self.query_date_range.to_placeholders(),
                **self._interval_placeholders(),
                "cross_join_select_query": cross_join_select_query,
            },
        )

    def _events_query(
        self, events_where_clause: ast.Expr, sample_value: ast.RatioExpr
    ) -> ast.SelectQuery | ast.SelectUnionQuery:
        return parse_select(
            """
                SELECT
                    timestamp as timestamp,
                    e.person_id AS actor_id
                FROM
                    events e
                SAMPLE {sample}
                WHERE {events_where_clause}
                GROUP BY
                    timestamp,
                    actor_id
            """,
            placeholders={
                "events_where_clause": events_where_clause,
                "sample": sample_value,
            },
        )

    def get_query_orchestrator(self, events_where_clause: ast.Expr, sample_value: ast.RatioExpr):
        events_query = self._events_query(events_where_clause, sample_value)
        inner_select = self._inner_select_query(events_query)
        parent_select = self._parent_select_query(inner_select)

        class QueryOrchestrator:
            events_query_builder: QueryAlternator
            inner_select_query_builder: QueryAlternator
            parent_select_query_builder: QueryAlternator

            def __init__(self):
                self.events_query_builder = QueryAlternator(events_query)
                self.inner_select_query_builder = QueryAlternator(inner_select)
                self.parent_select_query_builder = QueryAlternator(parent_select)

            def build(self):
                self.events_query_builder.build()
                self.inner_select_query_builder.build()
                self.parent_select_query_builder.build()

                return parent_select

        return QueryOrchestrator()
