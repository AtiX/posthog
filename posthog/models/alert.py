from django.db import models
from typing import Optional
from dataclasses import dataclass


@dataclass
class AbsoluteThreshold:
    lower: Optional[float] = None
    upper: Optional[float] = None


class AnomalyCondition:
    absolute_threshold: AbsoluteThreshold

    def __init__(self, absoluteThreshold: dict):
        self.absolute_threshold = AbsoluteThreshold(**absoluteThreshold)


class Alert(models.Model):
    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
    insight: models.ForeignKey = models.ForeignKey("posthog.Insight", on_delete=models.CASCADE)

    name: models.CharField = models.CharField(max_length=100)
    target_value: models.TextField = models.TextField()
    anomaly_condition: models.JSONField = models.JSONField(default=dict)
