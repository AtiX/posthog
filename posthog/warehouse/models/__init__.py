from .cluster import *
from .node import *
