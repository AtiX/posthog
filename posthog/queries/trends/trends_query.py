from posthog.models import Team, Filter
from posthog.queries.query_node_to_filter import query_node_to_filter
from posthog.queries.trends.trends import Trends
from posthog.schema import TrendsQuery, TrendsQueryResponse


def run_trends_query(
    team: Team,
    query: TrendsQuery,
) -> TrendsQueryResponse:
    data = query_node_to_filter(query)
    filter = Filter(data=data, team=team)
    trends_query = Trends()
    result = trends_query.run(filter, team)
    return TrendsQueryResponse(result=result, timezone=team.timezone)
