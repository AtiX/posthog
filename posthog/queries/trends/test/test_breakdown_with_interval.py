from datetime import datetime, timedelta
from typing import Dict

from freezegun import freeze_time

from posthog.models import Filter
from posthog.queries.trends.trends import Trends
from posthog.test.base import APIBaseTest, ClickhouseTestMixin, snapshot_clickhouse_queries
from posthog.test.test_journeys import journeys_for


@freeze_time("2022-11-09T21:14")
class TestBreakdownsWithInterval(ClickhouseTestMixin, APIBaseTest):
    """
    Two weeks of data, broken down by "some property" and interval "week"
    Not all broken down series have a value for each of the weeks

    freeze time to the 9th of November (a wednesday) and query for "-7d"
    so that the weeks are:
    - week 1: 30th of October - 5th of November (week containing another four of the seven days)
    - week 2: 6th - 12th of November (week containing three of the seven days)

    """

    def setUp(self):
        super().setUp()

        self.week_one = datetime(2022, 10, 30, 12, 1)  # sunday
        self.week_two = datetime(2022, 11, 6, 12, 1)  # sunday

        journey = {
            # count of 1 for each week from person 1 for "some property" = "a"
            "person1": [
                {
                    "event": "watched movie",
                    "properties": {"$some_property": "a"},
                    "timestamp": self.week_one,
                },
                {"event": "watched movie", "properties": {"$some_property": "a"}, "timestamp": self.week_two},
            ],
            # count of 1 for week 1 for "some property" = "b"
            "person2": [
                {
                    "event": "watched movie",
                    "properties": {"$some_property": "b"},
                    # why do I need to add 3 days before this result appears?!
                    "timestamp": self.week_one + timedelta(days=3),
                },
            ],
            # count of 1 for week 2 from person 3 for "some property" = "c"
            "person3": [
                {"event": "watched movie", "properties": {"$some_property": "c"}, "timestamp": self.week_two},
            ],
        }

        journeys_for(journey, team=self.team, create_people=True)

    def _run(self, extra: Dict = {}, events_extra: Dict = {}):
        response = Trends().run(
            Filter(
                data={
                    "events": [{"id": "watched movie", "name": "watched movie", "type": "events", **events_extra}],
                    "date_from": "-7d",
                    "interval": "week",
                    "breakdown": "$some_property",
                    "breakdown_type": "event",
                    **extra,
                }
            ),
            self.team,
        )
        return response

    @snapshot_clickhouse_queries
    def test_breakdown_by_some_property_by_week_interval(self):
        response = self._run({})

        assert [{"label": r["label"], "days": r["days"], "data": r["data"]} for r in response] == [
            {
                "data": [1.0, 1.0],
                "days": [
                    "2022-10-30",
                    "2022-11-06",
                ],
                "label": "watched movie - a",
            },
            {
                "data": [1.0, 0.0],
                "days": [
                    "2022-10-30",
                    "2022-11-06",
                ],
                "label": "watched movie - b",
            },
            {
                "data": [0.0, 1.0],
                "days": [
                    "2022-10-30",
                    "2022-11-06",
                ],
                "label": "watched movie - c",
            },
        ]
