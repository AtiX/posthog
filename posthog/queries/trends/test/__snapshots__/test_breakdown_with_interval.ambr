# name: TestBreakdownsWithInterval.test_breakdown_by_some_property_by_week_interval
  '
  
  SELECT groupArray(value)
  FROM
    (SELECT replaceRegexpAll(JSONExtractRaw(properties, '$some_property'), '^"|"$', '') AS value,
            count(*) as count
     FROM events e
     WHERE team_id = 2
       AND event = 'watched movie'
       AND toDateTime(timestamp, 'UTC') >= toDateTime('2022-11-02 00:00:00', 'UTC')
       AND toDateTime(timestamp, 'UTC') <= toDateTime('2022-11-09 23:59:59', 'UTC')
     GROUP BY value
     ORDER BY count DESC, value DESC
     LIMIT 25
     OFFSET 0)
  '
---
# name: TestBreakdownsWithInterval.test_breakdown_by_some_property_by_week_interval.1
  '
  
  SELECT groupArray(day_start) as date,
         groupArray(count) as data,
         breakdown_value
  FROM
    (SELECT SUM(total) as count,
            day_start,
            breakdown_value
     FROM
       (SELECT *
        FROM
          (SELECT toUInt16(0) AS total,
                  ticks.day_start as day_start,
                  breakdown_value
           FROM
             (SELECT toStartOfWeek(toDateTime('2022-11-09 23:59:59', 'UTC') - number * 604800) as day_start
              FROM numbers(2)
              UNION ALL SELECT toStartOfWeek(toDateTime('2022-11-02 00:00:00', 'UTC')) as day_start) as ticks
           CROSS JOIN
             (SELECT breakdown_value
              FROM
                (SELECT ['c', 'b', 'a'] as breakdown_value) ARRAY
              JOIN breakdown_value) as sec
           ORDER BY breakdown_value,
                    day_start
           UNION ALL SELECT count(*) as total,
                            toStartOfWeek(toTimeZone(toDateTime(timestamp, 'UTC'), 'UTC'), 0) as day_start,
                            replaceRegexpAll(JSONExtractRaw(properties, '$some_property'), '^"|"$', '') as breakdown_value
           FROM events e
           WHERE e.team_id = 2
             AND event = 'watched movie'
             AND toDateTime(timestamp, 'UTC') >= toStartOfWeek(toDateTime('2022-11-02 00:00:00', 'UTC'), 0)
             AND toDateTime(timestamp, 'UTC') <= toDateTime('2022-11-09 23:59:59', 'UTC')
             AND replaceRegexpAll(JSONExtractRaw(properties, '$some_property'), '^"|"$', '') in (['c', 'b', 'a'])
           GROUP BY day_start,
                    breakdown_value))
     GROUP BY day_start,
              breakdown_value
     ORDER BY breakdown_value,
              day_start)
  GROUP BY breakdown_value
  ORDER BY breakdown_value
  '
---
